src/QMCWaveFunctions/lcao/LCAOrbitalSet.h

```diff
@@ -60,11 +70,136 @@ namespace qmcplusplus
     virtual ~LCAOrbitalSet();
 
     SPOSet* makeClone() const;
+    ///test
+    void test()
+    {
+      app_log()<<"test function test function SDP \n";
+    }
+    /// create optimizable orbital rotation parameters
+    void buildOptVariables(std::vector<RealType>& input_params, bool params_supplied, std::vector<int> * data, const size_t& nel, std::vector<size_t>& C2node, const int& spin);
+
+    ///helper function to buildOptVariables
+    int build_occ_vec(std::vector<int> * data, const size_t& nel, const size_t& nmo, std::vector<int>* occ_vec);
+  
+    void evaluateDerivatives (ParticleSet& P,
```

Comment:
This function seems very heavy and seems only used for MSD optimization.
Is it possible to split it the math two levels. In the SPO level, only the evaluation of derivatives or some intermediate data shared by single SD and MSD and then in the single SD and MSD compute the final derivatives?

Reply:
I'm thinking of the case the spo_up and spo_dn is the same pointer for a restricted HF. This is very common for solid because double the spo adds more memory which we can't afford. LCAO code is less restrictive on memory. So the lazy solution is always building SPO twice even they are identical. In the case optimization, not taking into account spin up/down symmetry may cause severe issue because of spin contamination. This may needs a bit non-trivial modification of the current implementation. I will not block you PR because of this but it seems some thing we must solve.

Reply:
Any comment on this?